### ISSUE TITLE
HPX does not compile with G++ version prior to 9.3.

### ISSUE THREAD

**Issue by JiakunYan** [STEllAR-GROUP/hpx#6240]
## Expected Behavior

Compile successfully for any compilers supporting c++17.

## Actual Behavior

G++ 8.5 gave me an error `error: expected unqualified-id before '[' token`

## Steps to Reproduce the Problem

Just compile it with G++ 8.5.

## Specifications

  - HPX Version: the master branch `ad3ac76e159692355d4715f8b275c0a062b859af`
  - Platform (compiler, OS): G++ 8.5. This should apply to any version prior to g++9.3.

## Comment

This is not a fault of HPX, but G++, but this ticket can at least serve as an explanation for future users encountering this error. This is caused by a bug in G++ prior to 9.3. G++ doesn't parse correctly the `[[maybe_unused]]` attribute for the constructor's first argument. However, HPX uses it in `libs/core/executors/include/hpx/executors/service_executors.hpp` line 49.

```
explicit service_executor([[maybe_unused]] hpx::util::io_service_pool* pool) noexcept
```

Reference: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=81429.

**Comment by hkaiser** [comment]
We can make this attribute conditional, though.

**Comment by Johan511** [comment]
Do we look into replacing maybe_unused with a #define, possibly something like HPX_UNUSED which falls back to __attribute__((unused)) for older GCC versions?

**Comment by hkaiser** [comment]
We do have `HPX_UNUSED`, however its use is different from `[[maybe_unused]]`. So instead of writing:
```
void foo([[maybe_unused]] int bar) {}
```
one would have to write:
```
void foo(int bar) { HPX_UNUSED(bar); }
```
We should make those changes only singular places when necessary, though.

**Comment by Johan511** [comment]
So everywhere we use [[maybe_unused]] should be replaced with HPX_UNUSED

or we could have another #define for [[maybe_unused]] which falls back to __attribute((unused))

Any comments on what you believe is better?

**Comment by hkaiser** [comment]
> So everywhere we use [[maybe_unused]] should be replaced with HPX_UNUSED

No! Everywhere possible where we currently use `HPX_UNUSED` we should use `[[maybe_unused]]` instead.

> or we could have another #define for [[maybe_unused]] which falls back to __attribute((unused))

How many spots would that be used in? If it's just a handful, I think it's not worth the trouble.

**Comment by hkaiser** [comment]
Closing this as 'Won't fix'.
